### PR TITLE
Add `SkyProgressIndicatorMessage`

### PR DESCRIPTION
**Pull request by Blackbaud-SteveBrush**

### CODE REVIEW

File: src/app/public/modules/progress-indicator/progress-indicator.component.ts [hunk 1]
```diff
@@ -1,110 +1,138 @@
...
   @Input()
-  public get displayMode(): SkyProgressIndicatorDisplayMode {
-    return this._displayMode || SkyProgressIndicatorDisplayMode.Vertical;
-  }
   public set displayMode(value: SkyProgressIndicatorDisplayMode) {
     this._displayMode = value;
   }
 
+  public get displayMode(): SkyProgressIndicatorDisplayMode {
+    if (this._displayMode === undefined) {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/progress-indicator/progress-indicator.component.ts:49]: Why did you remove the old short circuit?

File: src/app/public/modules/progress-indicator/progress-indicator.component.ts [hunk 2]
```diff
@@ -1,110 +1,138 @@
...
-  private activeIndex = 0;
+  private get activeIndex(): number {
+    return this._activeIndex || 0;
+  }
+
+  private set activeIndex(value: number) {
+    const lastIndex = this.progressItems.length - 1;
+
+    if (value > lastIndex) {
+      value = lastIndex;
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/progress-indicator/progress-indicator.component.ts]: Is it good practice to modify the value coming in? Instead, we could just assign the private variable inside these conditionals like so...
```
if (value > lastIndex) {
  this._activeIndex = lastIndex;
}
if (value < 0) {
  this._activeIndex = 0;
}
```

File: src/app/public/modules/progress-indicator/progress-indicator-title/progress-indicator-title.component.html [hunk 3]
```diff
@@ -1,4 +1,6 @@
-<h3
-  class="sky-section-heading sky-progress-indicator-title">
+<div
+  class="sky-progress-indicator-title sky-section-heading"
+  role="heading"
+>
   <ng-content></ng-content>
-</h3>
+</div>
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/progress-indicator/progress-indicator-title/progress-indicator-title.component.html:6]: Why did you change this from an `h3` to a `div` ?

File: src/app/public/modules/progress-indicator/progress-indicator.component.spec.ts [hunk 4]
```diff
@@ -181,29 +159,29 @@ describe('Progress indicator component', function () {
...
       fixture.detectChanges();
-      componentInstance.messageStream.next(SkyProgressIndicatorMessageType.Progress);
+      componentInstance.sendMessage(SkyProgressIndicatorMessageType.Progress);
       fixture.detectChanges();
-      componentInstance.messageStream.next(SkyProgressIndicatorMessageType.Progress);
+      componentInstance.sendMessage(SkyProgressIndicatorMessageType.Progress);
       fixture.detectChanges();
 
-      expect(componentInstance.activeIndex).toBe(3);
-      expect(componentInstance.progressItems.last.isActive).toBeFalsy();
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/progress-indicator/progress-indicator.component.spec.ts:192]: I'm confused... How did this test pass before?

**Review by Blackbaud-AlexKingman — COMMENTED** [review]
Few questions for you @Blackbaud-SteveBrush .

File: src/app/public/modules/progress-indicator/progress-indicator.component.ts [hunk 5]
```diff
@@ -216,4 +260,62 @@ export class SkyProgressIndicatorComponent implements AfterContentInit, OnDestro
...
+    let type: SkyProgressIndicatorMessageType;
+
+    // Prints a deprecation warning if the consumer provides only `SkyProgressIndicatorMessageType`.
+    if (value.type === undefined) {
+      console.warn(
+        '[Deprecation warning] The progress indicator component\'s `messageStream` input was ' +
+        'set to `Subject<SkyProgressIndicatorMessageType>`. This is a ' +
+        'deprecated type and will be removed in the next major version release. ' +
+        'Instead, set the `messageStream` input to `Subject<SkyProgressIndicatorMessage>`.'
+      );
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/progress-indicator/progress-indicator.component.ts:278]: From @blackbaud-johnly - we should be writing these in the active voice. For example:

[Deprecation warning] The progress indicator component's `messageStream` input is set to `Subject<SkyProgressIndicatorMessageType>`. We will remove this deprecated type in the next major version release. Instead, set the `messageStream` input to `Subject<SkyProgressIndicatorMessage>`.

File: src/app/public/modules/progress-indicator/progress-indicator.component.ts (hunk 1 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/progress-indicator/progress-indicator.component.ts:49]: Because the first value of `SkyProgressIndicatorDisplayMode` is zero (it's a numeric `enum`), and the original short circuit evaluated `0` to false, we need to check for `undefined` specifically, since that was the spirit of the check to begin with.

File: src/app/public/modules/progress-indicator/progress-indicator.component.ts (hunk 2 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/progress-indicator/progress-indicator.component.ts]: For a setter, I personally feel that the actual field (`_activeIndex`) should be set only once. I do agree that we shouldn't be modifying the argument; I'll update the method to set the `value` to a `newIndex` variable, instead.

File: src/app/public/modules/progress-indicator/progress-indicator.component.ts (hunk 5 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/progress-indicator/progress-indicator.component.ts:278]: Good catch.

File: src/app/public/modules/progress-indicator/progress-indicator-title/progress-indicator-title.component.html (hunk 3 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/progress-indicator/progress-indicator-title/progress-indicator-title.component.html:6]: Matt Gregg mentioned that component "headings" should not use heading HTML tags because this can throw off screen readers (it's still okay to use heading tags in general markup such as above paragraphs, etc.). For our components, we should be using `div` with a `role` of "heading".

File: src/app/public/modules/progress-indicator/progress-indicator.component.spec.ts (hunk 4 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/progress-indicator/progress-indicator.component.spec.ts:192]: It passed because the logic of the steps was incorrect. The final step has an index of `2` because there's only three items. The final step should also be marked as `active`. Let me know if I'm missing something obvious!

File: src/app/public/modules/progress-indicator/progress-indicator.component.ts (hunk 1 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/progress-indicator/progress-indicator.component.ts:49]: Makes sense. Thanks!

File: src/app/public/modules/progress-indicator/progress-indicator-title/progress-indicator-title.component.html (hunk 3 above)
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/progress-indicator/progress-indicator-title/progress-indicator-title.component.html:6]: Hm... I'm interested in hearing more about this. In the meantime, I'll take your word and we can move forward!

**Review by Blackbaud-AlexKingman — APPROVED** [review]
LGTM! Thanks for splitting this up.
